lib/_included_packages/plexnet/videosession.py

```diff
@@ -0,0 +1,415 @@
+
+
+class MediaDetails:
+    """
+    Gathers attributes from a MediaContainer
+    """
+    details = None
+    attribute_map = {
+        "container": "part.attrib_container",
+        "partDecision": "part.decision",
+        "videoResolution": "media.videoResolution",
+        "videoBitrate": "video_stream.bitrate",
+        "videoCodec": "video_stream.codec",
+        "videoStreamDecision": "video_stream.decision",
+        "transcodeVideoDecision": "transcode_session.videoDecision",
+        "transcodeHWEncoding": "transcode_session.transcodeHwEncoding",
+        "audioCodec": "audio_stream.codec",
+        "audioBitrate": "audio_stream.bitrate",
+        "audioChannels": "audio_stream.channels",
+        "audioStreamDecision": "audio_stream.decision",
+        "subtitleCodec": "subtitle_stream.codec",
+        "subtitleStreamDecision": "subtitle_stream.decision",
+        "subtitleLocation": "subtitle_stream.location",
+        "subtitleBurn": "subtitle_stream.burn",
+    }
+
+    def __init__(self, *args, **kwargs):
+        self.details = self.findMediaDetails(*args, **kwargs)
+
+    def attributesFromInstance(self, map, reference_data):
+        # gather attribute values
+        final_data = {}
+        for attribute, dataPath in map.iteritems():
+            objName, attribName = dataPath.split(".")
+            if objName in reference_data:
+                obj = reference_data[objName]
+                final_data[attribute] = getattr(obj, attribName, None) or None
+        return final_data
+
+    def findMediaDetails(self, mediaContainer, mediaChoice, transcodeSession=None):
+        """
+
+        """
+        reference_data = {
+            "media": None,
+            "part": None,
+            "video_stream": None,
+            "audio_stream": None,
+            "subtitle_stream": None,
+            "transcode_session": transcodeSession or mediaContainer.transcodeSession
+        }
+
+        # We can't use mediaChoice here directly, because it doesn't necessarily hold the newest data (in case of
+        # an actual MediaContainer from the Session endpoint that data is king)
+        # Instead find the media/part/streams which were selected and are held in MediaChoice inside the current
+        # mediaContainer
+        for media in mediaContainer.media:
+            if media.id == mediaChoice.media.id:
+                reference_data["media"] = media
+
+                for part in media.parts:
+                    if part.id == mediaChoice.part.id:
+                        reference_data["part"] = part
+
+                        for stream in part.streams:
+                            if stream.id == mediaChoice.videoStream.id:
+                                reference_data["video_stream"] = stream
+                            elif stream.id == mediaChoice.audioStream.id:
+                                reference_data["audio_stream"] = stream
+                            elif stream.id == mediaChoice.subtitleStream.id:
```

Comment:
Before executing `mediaChoice.subtitleStream.id`, you should check to make sure the subtitle stream exists (it doesn't always) otherwise this will throw.  So this line is better as `elif mediaChoice.subtitleStream != None and stream.id == mediaChoice.subtitleStream.id:`  Similar should be done with the other streams just to make sure.